src/asynch/runner.rs

```diff
@@ -265,53 +362,235 @@ impl<'d, AT: AtatClient, C: CellularConfig, const URC_CAPACITY: usize>
             .await?;
         self.at
-            .send(SetModuleFunctionality {
+            .send(&SetModuleFunctionality {
                 fun: Functionality::Full,
                 rst: Some(ResetMode::DontReset),
             })
             .await?;
 
-        Err(Error::Busy)
+        Ok(())
     }
 
+    /// Reset the module by driving it's `RESET_N` pin low for 50 ms
+    ///
+    /// **NOTE** This function will reset NVM settings!
     pub async fn reset(&mut self) -> Result<(), Error> {
         warn!("Hard resetting Ublox Cellular Module");
         if let Some(pin) = self.config.reset_pin() {
             pin.set_low().ok();
             Timer::after(reset_time()).await;
             pin.set_high().ok();
             Timer::after(boot_time()).await;
-            self.ch.set_power_state(PowerState::PowerUp);
+            self.is_alive().await?;
         } else {
             warn!("No reset pin configured");
         }
         Ok(())
     }
 
-    pub async fn restart(&mut self, store: bool) -> Result<(), Error> {
+    /// Perform at full factory reset of the module, clearing all NVM sectors in the process
+    pub async fn factory_reset(&mut self) -> Result<(), Error> {
+        self.at
+            .send(&crate::command::system_features::SetFactoryConfiguration {
+                fs_op: crate::command::system_features::types::FSFactoryRestoreType::AllFiles,
+                nvm_op:
+                    crate::command::system_features::types::NVMFactoryRestoreType::NVMFlashSectors,
+            })
+            .await?;
+
+        info!("Successfully factory reset modem!");
+
+        if self.soft_reset(true).await.is_err() {
+            self.reset().await?;
+        }
+
         Ok(())
     }
 
+    /// Reset the module by sending AT CFUN command
+    pub async fn soft_reset(&mut self, sim_reset: bool) -> Result<(), Error> {
+        trace!(
+            "Attempting to soft reset of the modem with sim reset: {}.",
+            sim_reset
+        );
+
+        let fun = if sim_reset {
+            Functionality::SilentResetWithSimReset
+        } else {
+            Functionality::SilentReset
+        };
+
+        match self
+            .at
+            .send(&SetModuleFunctionality {
+                fun,
+                // SARA-R5 This parameter can be used only when <fun> is 1, 4 or 19
+                #[cfg(feature = "sara-r5")]
+                rst: None,
+                #[cfg(not(feature = "sara-r5"))]
+                rst: Some(ResetMode::DontReset),
+            })
+            .await
+        {
+            Ok(_) => {
+                info!("Successfully soft reset modem!");
+                Ok(())
+            }
+            Err(err) => {
+                error!("Failed to soft reset modem: {:?}", err);
+                Err(Error::Atat(err))
+            }
+        }
+    }
+
+    // checks alive status continuiously until it is alive
+    async fn check_is_alive_loop(&mut self) -> bool {
+        loop {
```

Comment:
Hmm.. Think I would prefer if this had either a retry counter or a timeout..? This seems like something that can get stuck for infinity?

Reply:
It should have a timeout in the transition match. Extracted it to be less code in the match.